index.html

```diff
@@ -2209,218 +2270,232 @@ <h3><dfn>Writing or pushing content</dfn></h3>
-                            The error here gives an indication that the operation
-                            needs to be repeated. Otherwise the user may think the
-                            operation succeeded on all connected <a>NFC adapter</a>s.
-                          </p>
-                        </li>
-                        <li>
-                          When the transfer has completed, clear |output|
-                          associated with |target|, resolve
-                          |p|.
-                        </li>
-                      </ol>
-                    </li>
-                  </ol>
-                </dl>
+                Verify the following conditions:
+                <ul>
+                  <li>
+                    if |device| is an <a>NFC tag</a>, |target| is
+                    "`tag`" or "`any`".
+                  </li>
+                  <li>
+                    if |device| is an <a>NFC peer</a>, |target| is
+                    "`peer`" or "`any`".
+                  </li>
+                  <li>
+                    <a href="#nfc-is-suspended">NFC is not suspended</a>.
+                  </li>
+                </ul>
               </li>
               <li>
-                [=promise/React=] to |p|:
+                In case of success, run the following steps:
                 <ol>
                   <li>
-                    If |p| was settled (fulfilled or rejected), then:
-                    <ol>
+                    Stop |writer|.[[\PushTimer]] if active.
+                  </li>
+                  <li>
+                    If |device| is an <a>NFC tag</a>,
+                    <ul>
+                      <li>Read the tag.</li>
+                      <li>
+                        If |options|'s ignoreRead is not equal to
+                        `true`, run the
+                        <a href="#steps-receiving">receiving steps</a>.
+                      </li>
                       <li>
-                        Clears the <a>pending push tuple</a>.
+                        Let |idHost| be the <a>record identifier host</a> of the
+                        tag.
                       </li>
                       <li>
-                        Removes <a>async write handlers</a>.
+                        If |idHost| is `null`, or different than the
+                        <a data-cite="url#concept-host-serializer">
+                        serialized host</a> of the <a>current settings object</a>'s
+                        origin, and the <a>obtain push permission</a> steps return
+                        `false`, then reject |p| with {{"NotAllowedError"}}
+                        {{DOMException}} and abort these steps.
                       </li>
-                    </ol>
+                    </ul>
+                  </li>
+                  <li>
+                    Let |output:NDEFMessage| be |writer|.[[\PushMessage]].
```

Comment:
It seems to me that today we have the (p, writer) tuple because of the async write handlers, but you are getting rid of that here, so with that we don't really need to store the writer at all.

Also all the internal slots you are adding to writer, are actually globals so make more sense on the NFC state record.

Just add "PushPromise, PushOptions, PushMessage, PushTimer" there.

Then we add react steps to the promise to clear PushOptions, PushMessage and PushTimer, so that things are always consistent.

If a new push is scheduled, just check whether PushPromise exists, if it does, reject it which will synchnonously clean everything up, then add the new ones.

Reply:
Hmm, I am fine with that. I didn't want to go that far to remove the tuple, but actually it would simplify things further.
I can do that tomorrow.

Reply:
And thanks for the review.